dask/array/routines.py

```diff
@@ -1227,3 +1228,53 @@ def insert(arr, obj, values, axis):
     interleaved = list(interleave([split_arr, split_values]))
     interleaved = [i for i in interleaved if i.nbytes]
     return concatenate(interleaved, axis=axis)
+
+
+@wraps(np.average)
+def average(a, axis=None, weights=None, returned=False):
+    # This was minimally modified from numpy.average
+    # See numpy license at https://github.com/numpy/numpy/blob/master/LICENSE.txt
+    # or NUMPY_LICENSE.txt within this directory
+    a = asanyarray(a)
+
+    if weights is None:
+        avg = a.mean(axis)
+        scl = avg.dtype.type(a.size / avg.size)
+    else:
+        wgt = asanyarray(weights)
+
+        if issubclass(a.dtype.type, (np.integer, np.bool_)):
+            result_dtype = result_type(a.dtype, wgt.dtype, 'f8')
+        else:
+            result_dtype = result_type(a.dtype, wgt.dtype)
+
+        # Sanity checks
+        if a.shape != wgt.shape:
+            if axis is None:
+                raise TypeError(
+                    "Axis must be specified when shapes of a and weights "
+                    "differ.")
+            if wgt.ndim != 1:
+                raise TypeError(
+                    "1D weights expected when shapes of a and weights differ.")
+            if wgt.shape[0] != a.shape[axis]:
+                raise ValueError(
+                    "Length of weights not compatible with specified axis.")
+
+            # setup wgt to broadcast along axis
+            wgt = broadcast_to(wgt, (a.ndim - 1) * (1,) + wgt.shape)
+            wgt = wgt.swapaxes(-1, axis)
+
+        scl = wgt.sum(axis=axis, dtype=result_dtype)
+        if (scl == 0.0).any():
```

Comment:
Calling the `any()` method here triggers an array  `compute`. Still thinking of a way around this.

Reply:
It would probably be better to not worry about this check, and let the result of dividing by zero be NaN. NumPy will already issue a warning (by default) or raise for this, based on floating point error handling setting: https://docs.scipy.org/doc/numpy-1.14.0/reference/routines.err.html

Reply:
Thanks for the review @shoyer, that's a good point! I'll remove the check and swap the `ZeroDivisionError` test for a `RuntimeWarning` test.